### PR TITLE
Add cryptotokenPrivate and device support (hid and usb) for U2F

### PR DESCRIPTION
**Pull request by evq**
This PR contains the following changes required to enable u2f support:
1. Instantiate the chrome device client to enable the existing `chrome.hid` and `chrome.usb` apis
2. Implement `chrome.windows.get` which is used by the cryptotoken extension
3. Build and register the `chrome.cryptotokenPrivate` extension api
4. Allow loading of component extensions from pak resources via the brave extension api

I have tested these changes on linux, mac and windows.

### CODE REVIEW

File: atom/common/api/resources/windows_bindings.js [hunk 1]
```diff
@@ -17,7 +17,21 @@ binding.registerCustomHook(function (bindingsAPI, extensionId) {
   })
 
   apiFunctions.setHandleRequest('get', function (windowId, getInfo, cb) {
-    console.warn('chrome.windows.get is not supported yet')
+    if (arguments.length === 1) {
```
Comment (bridiver) [inline atom/common/api/resources/windows_bindings.js]: you actually don't need all of these checks because`setHandleRequest` normalizes the arguments

File: atom/common/api/resources/windows_bindings.js [hunk 2]
```diff
@@ -17,7 +17,21 @@ binding.registerCustomHook(function (bindingsAPI, extensionId) {
...
+      cb = arguments[0]
+    } else if (arguments.length == 2) {
+      getInfo = arguments[0]
+      cb = arguments[1]
+    } else {
+      windowId = arguments[0]
+      getInfo = arguments[1]
+      cb = arguments[2]
+    }
+    var responseId = ++id
```
Comment (bridiver) [inline atom/common/api/resources/windows_bindings.js]: we've moved away from using `++id` because it can produce duplicate requests across multiple windows. See `tabs_bindings.js` for usage of `ipc.guid()`

File: BUILD.gn [hunk 3]
```diff
@@ -121,6 +122,7 @@ repack("packed_extra_resources") {
     ":atom_resources",
     ":brave_resources",
     "app:brave_strings",
+    "//chrome/browser/resources:component_extension_resources_grit",
```
Comment (bridiver) [inline BUILD.gn]: shouldn't think be `//chrome/browser/resources:component_extension_resources`? I don't see a target for `component_extension_resources_grit`

File: atom/browser/extensions/atom_component_extension_resource_manager.cc [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      kComponentExtensionResourcesSize);
+  AddComponentResourceEntries(
+      kExtraComponentExtensionResources,
+      arraysize(kExtraComponentExtensionResources));
+}
+
+AtomComponentExtensionResourceManager::
+~AtomComponentExtensionResourceManager() {}
+
+bool AtomComponentExtensionResourceManager::IsComponentExtensionResource(
```
Comment (bridiver) [inline atom/browser/extensions/atom_component_extension_resource_manager.cc]: most of this code looks identical to `ChromeComponentExtensionResourceManager`. We want to avoid copying chrome code so can you subclass instead?

Comment (evq) [inline atom/browser/extensions/atom_component_extension_resource_manager.cc]: I don't think I can subclass since `AddComponentResourceEntries` and `path_to_resource_id_` are private

File: chromium_src/BUILD.gn [hunk 5]
```diff
@@ -400,10 +404,18 @@ source_set("browser") {
...
         "//chrome/browser/extensions/api/tabs/tabs_constants.h",
+        "//chrome/browser/extensions/api/tabs/windows_util.cc",
+        "//chrome/browser/extensions/api/tabs/windows_util.h",
         "//chrome/browser/extensions/chrome_extension_api_frame_id_map_helper.cc",
         "//chrome/browser/extensions/chrome_extension_api_frame_id_map_helper.h",
+        "//chrome/browser/extensions/chrome_extension_function_details.cc",
+        "//chrome/browser/extensions/chrome_extension_function_details.h",
         "//chrome/browser/extensions/event_router_forwarder.cc",
         "//chrome/browser/extensions/event_router_forwarder.h",
+        "//chrome/browser/extensions/window_controller.cc",
```
Comment (bridiver) [inline chromium_src/BUILD.gn]: this looks like a potential problem to me because along with windows_util above. None of these methods will work correctly without code so anything that uses them will get the wrong result. I think we need to look at the usage and possibly override with our own implementation

File: lib/browser/api/extensions.js [hunk 6]
```diff
@@ -603,6 +603,28 @@ var windowInfo = function (win, populateTabs) {
   }
 }
 
+ipcMain.on('chrome-windows-get', function (evt, responseId, windowId, getInfo) {
```
Comment (bridiver) [inline lib/browser/api/extensions.js]: is this a rebase issue? Doesn't seem related

Comment (evq) [inline lib/browser/api/extensions.js]: `chrome.windows.get` is needed by the `cryptotoken` extension, thus I implemented it

**Comment by alexwykoff** [comment]
What are the next steps @evq ?

**Comment by evq** [comment]
@alexwykoff 
I have some local changes to address the requested changes which I was working on pre-Mercury, I need to rebase and finish those off. Barring major stumbling blocks during rebase, it shouldn't take much time at all once I can get back at it. Would really like to target the end of next week.

**Comment by evq** [comment]
@bridiver apologies for the long hiatus on this one... finally got some time to rebase, finish addressing your comments and test!

the primary change avoids copying a bunch of code from `ChromeComponentExtensionResourceManager` - I am now using it directly and have removed the small shim `AtomComponentExtensionResourceManager`. I've created a helper method to enable loading the extension manifest from the resources bundle.

File: atom/browser/BUILD.gn [hunk 7]
```diff
@@ -310,6 +310,8 @@ source_set("browser") {
       "extensions/api/atom_extensions_api_client.h",
       "extensions/atom_browser_client_extensions_part.cc",
       "extensions/atom_browser_client_extensions_part.h",
+      "extensions/atom_component_extensions.cc",
```
Comment (bridiver) [inline atom/browser/BUILD.gn]: since this uses brave_resources.h we need to add a dep for `brave_resources` here too

File: atom/browser/BUILD.gn [hunk 8]
```diff
@@ -306,6 +308,8 @@ source_set("browser") {
   }
 
   if (enable_extensions) {
+    data_deps += [ "//electron:brave_resources" ]
```
Comment (evq) [inline atom/browser/BUILD.gn]: is this the right way to address your previous comment re: the missing `brave_resources.h` dependency? wasn't sure since it is generated and I'm not particularly familiar with GN

File: brave/browser/api/brave_api_extension.cc [hunk 9]
```diff
@@ -98,10 +107,17 @@ scoped_refptr<extensions::Extension> LoadExtension(const base::FilePath& path,
     return NULL;
 
   std::vector<extensions::InstallWarning> warnings;
-  if (!extensions::file_util::ValidateExtension(extension.get(),
-                                                error,
-                                                &warnings))
-    return NULL;
+
+  int resource_id;
+  if (!IsComponentExtension(path, &resource_id)) {
```
Comment (bridiver) [inline brave/browser/api/brave_api_extension.cc]: can we use IsComponentExtensionResource here instead? `IsComponentExtension` is a bit misleading because we have component extensions that are not loaded from resource paks

File: muon/browser/muon_browser_process_impl.cc [hunk 10]
```diff
@@ -19,6 +20,8 @@ MuonBrowserProcessImpl::MuonBrowserProcessImpl(
       const base::CommandLine& command_line) :
     BrowserProcessImpl(local_state_task_runner, command_line) {
   g_browser_process = this;
+
+  device_client_.reset(new ChromeDeviceClient);
```
Comment (bridiver) [inline muon/browser/muon_browser_process_impl.cc]: this should actually go in chromium_src browser_process_impl.cc
the muon subclass is for muon-specific stuff
